### ISSUE TITLE
[flash_ctrl] Patch up corner cases in DV after merging the scrambling module

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#22089]
I still seem to be running into issues with error case handling. 
For some reason the shared module seems to be causing issues with
```
flash_ctrl_oversize_error
flash_ctrl_ro_derr
flash_ctrl_rw_derr
flash_ctrl_integrity
flash_ctrl_rw_serr
```

I am a bit hesitant merging this without fully understanding these errors, and it is a critical block for the system.

_Originally posted by @msfschaffner in https://github.com/lowRISC/opentitan/issues/22045#issuecomment-2000916780_

**Comment by msfschaffner** [comment]
This should probably be looked at in order to close down #21012

**Comment by matutem** [comment]
I see integrity errors flagged by flash_ctrl_oversize_error with seed 101526864083573723368910117507542387818288841624698717015324241278126026539731.

The flash_phy_rd block generates integrity errors around time 51787ps in https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/flash_ctrl/rtl/flash_phy_rd.sv;drc=8046c2896fa50aaf3a186a7ce8c0570db9f99eaf;l=677, but I see the top 4 bits in data_out_muxed don't change with the data contents. There is something weird in https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/flash_ctrl/rtl/flash_phy_rd.sv;drc=8046c2896fa50aaf3a186a7ce8c0570db9f99eaf;l=586: forward == 0 and hint_descram == 1, the integrity bits are from fifo_data, which is constant, while the actual data is from descramble_data_i and mask, which change.

**Comment by msfschaffner** [comment]
There were two exclusive instances of the scrambling module - one per bank.
We changed that and shared one scrambling module among both banks.
Specifically, there is a cipher datapath and a galois multiplier inside the scrambling module. Both operate independently, hence we used two RR arbiters that independently arbitrate the bank requests for these two sub-datapaths.

**Comment by matutem** [comment]
I think the problem is that the scrambler is missing creating a calc response for bank 0, so the key sequence gets messed up. The waves show both banks requesting masks at 51267, and there is no ack for bank 0. Once the mask is wrong everything fails.

![Screenshot from 2024-03-21 11-43-28](https://github.com/lowRISC/opentitan/assets/80294896/db1f7604-7501-4953-9d26-2384465d02f1)

**Comment by matutem** [comment]
One more detail: the request info needed by the mask computation is held in `rd_attrs`, but its lifetime is only while the request is waiting for data from storage: https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/flash_ctrl/rtl/flash_phy_rd.sv;drc=8046c2896fa50aaf3a186a7ce8c0570db9f99eaf;l=353. If we intend to keep requests pipelined we will also need to pipeline rd_attrs.

**Comment by msfschaffner** [comment]
I looked a bit at this. It appears that the req/ack interface to the scrambling module is not properly observed (the input data can change while a req has not been acked yet), and there is an implicit assumption that the scrambling module can immediately accept an incoming address for calculating the mask.
We could try and fix the back pressure logic at the input and/or add a holding reg, but it looks like it would have to be fixed for both host and controller requests. I worry that we may run into more implicit assumptions and introduce bugs at this point, hence my suggestion is to revert the change & reopen the tracker issue if we manage to save enough area elsewhere.

**Comment by vogelpi** [comment]
Thanks for your investigations @matutem and your input @msfschaffner ! For now, I've created a draft PR here #22191 to revert the change. I will now have a look as well how this could be fixed for good, as I believe we should keep the change if possible. If fixing is indeed to complicated, we can merge the reversion PR for the next RC.

**Comment by vogelpi** [comment]
I've had a look now at two different cases. The one you investigated above @matutem and another one (`util/dvsim/dvsim.py hw/ip/flash_ctrl/dv/flash_ctrl_sim_cfg.hjson -t vcs -i flash_ctrl_ro_derr --purge --reseed 1 --seed 28896645993018003030575529967580815634373933010200160736792270883106218439223`). In both cases, problems seem to start when the shared scrambling module gets a request from both banks in the same cycle. And in both cases, the arbiter gives priority to the request from Bank 1.

To me, the issue here seems to be that the scoreboard doesn't know which bank won the arbitration. I've been searching but couldn't find where the scoreboard takes the requests out of the read queue. I assume there would be two queues, one for each bank?

I tried hacking the scrambling primitive to always give higher priority to requests coming from Bank 0, but the problem with my hacks were that the arbiter then changes switches inputs of the Galois multiplier whenever a request from Bank 0 comes in while it's actually busy with Bank 1.

@matutem , do you understand where DV takes the requests out of the queues for checking? Could we change this part to push the request back in and try the next one in case of failures? I really think there is no fundamental RTL issue here as just the error tests fail. And they fail only when both banks request a read at the very same time, and only if priority is given to Bank 1.

**Comment by matutem** [comment]
The problem is not in DV, but in the RTL. The integrity error should not occur, but the hardware is using the scramble module incorrectly and a descramble request is missed entirely so the data from the flash read ends up using the wrong mask. If the hardware works correctly DV will show no problem, with no need to change it.

**Comment by matutem** [comment]
I found the simplest way to diagnose this was by running the same test with the same build-seed on a the flash_ctrl model prior to the problematic change. Comparing the waves for the scramble behavior in both will show the bug.

**Comment by vogelpi** [comment]
I am almost certain the bug is in the DV and not in the RTL: At time zero,
DV inserts de ECC error at several addresses including the address of
interest. With higher verbosity, you can see this in the run log.

In the case I analyzed, a first ECC error is correctly reported by the RTL
and DV. The second one is not expected in the scoreboard even though DV did
insert it initially (and the RTL correctly generates it).

Anyway, I acknowledge that we won't resolve this before the next RC is
tagged, so let's move ahead and merge the PR reverting the RTL. We can
continue investigating afterwards @andreaskurth.

**Comment by vogelpi** [comment]
Just had another look and managed to hack the prim_arbiter_tree to always prioritize Bank 0. The simulation then continues a bit further but again terminates when both banks make a request in the same clock cycle and one of the banks produces an error. Again, the error is produced for an address where the simulation previously inserted an error. So the error should be expected. It's not at time zero but I can see many messages like this:
```
UVM_INFO @ 10799.2 ns: (flash_ctrl_otf_base_vseq.sv:767) uvm_test_top.env.virtual_sequencer [direct_read] bank:0 tl_addr:0x6ef24, num: 22
UVM_INFO @ 10799.2 ns: (flash_ctrl_env_cfg.sv:258) [get_region] page:221 --> region:8
UVM_INFO @ 10799.2 ns: (flash_ctrl_otf_base_vseq.sv:817) uvm_test_top.env.virtual_sequencer [direct_read] ierr_created[1]:0  derr_is_set:0 exists:0
```
The final message is then:
```
UVM_ERROR @ 12819.4 ns: (cip_base_scoreboard.sv:481) [uvm_test_top.env.scoreboard] Check failed item.d_error == exp_d_error (1 [0x1] vs 0 [0x0]) On interface flash_ctrl_efla
sh_reg_block, TL item: req: (cip_tl_seq_item@91742) { a_addr: 'h6ef24  a_data: 'h0  a_mask: 'hf  a_size: 'h2  a_param: 'h0  a_source: 'h12  a_opcode: 'h4  a_user: 'h279aa  d_param: 'h0  d_s
ource: 'h12  d_data: 'hffffffff  d_size: 'h2  d_opcode: 'h1  d_error: 'h1  d_sink: 'h0  d_user: 'heaa  a_source_is_overridden: 'h0  a_valid_delay: 'h0  d_valid_delay: 'h0  a_valid_len: 'h0 
 d_valid_len: 'h0  req_abort_after_a_valid_len: 'h0  rsp_abort_after_d_valid_len: 'h0  req_completed: 'h0  rsp_completed: 'h0  tl_intg_err_type: TlIntgErrNone  max_ecc_errors: 'h3  }
                , unmapped_err: 0, mem_access_err: 0, bus_intg_err: 0, byte_wr_err: 0, csr_size_err: 0, tl_item_err: 0, write_w_instr_type_err: 0, cfg.tl_mem_access_gated: 0 ecc_err: 0
```
So the scoreboard says it sees an error which it didn't expect. But my understanding is that the previous message says, there should be one.

BTW: I've seen multiple occasions now where both banks push a request in the same cycle (without an error being produced and expected). So, I am pretty sure the sharing of the scrambling unit works but not the tracking in case of errors.

**Comment by vogelpi** [comment]
I've now done the same @matutem and can confirm your assessment. The GF multiplier sometimes gets the wrong input. We need to fix this in the RTL not in the DV. Apologies for stating this was a DV issue @matutem . What I wasn't aware of is the fact that single-bit ECC errors (which some these tests insert) get corrected in Flash upon read out without flagging a TL-UL error (they generate alerts though).

I am now looking into if and how we could fix this in the RTL without reverting the change.

**Comment by vogelpi** [comment]
I've now filed PR to which hopefully fixes these issues (full regression is currently running). What I ended up doing is close to what @matutem suggested: If the current request requires descrambling, new requests are not accepted from the host interface nor forwarded to the Flash primitive until the mask calculation is done.

The controller interface doesn't need additional fixing as the issue is not present there: controller requests are only accepted if the read pipeline is completely empty. There is no way of dropping a controller request (or the relevant read attributes) before the next one is sent in.
